Media / Attachments REST API endpoint: update attachments controller to support flip and to customize attachment fields.

ramonjd opened the pull request:
## What?

This PR is the first step to enhance media editor capabilities as described in #55238 [Phase 3: Collaboration > Media Library](https://github.com/WordPress/gutenberg/issues/55238).

It adds the following functionality:

- the ability to flip an image horizontally and vertically
- the ability to send arguments to update the new image's `caption`, `description`, and `title`, `post` and `alt_text` fields.


Resolves https://github.com/WordPress/gutenberg/issues/71863

Core trac ticket: https://core.trac.wordpress.org/ticket/64035


## Why?

As outlined in https://github.com/WordPress/gutenberg/issues/55238, the extended functionality will allow folks to start experimenting with a redesigned media library.

It will also allow flip controls in the existing image block toolbar.

## How?

- Adds new controller and override existing endpoint
- Avoids conflicts with existing media endpoint extension for the media processing experiment by renaming it to `Gutenberg_Media_Processing_REST_Attachments_Controller`.

## Testing Instructions

For the new tests:

`npm run test:unit:php:base -- --filter Gutenberg_REST_Attachments_Controller_Test`

Ensure that all tests pass (after updates to media processing tests)

`npm run test:unit:php`

### Manual testing

Check that existing use of the `media/{$id}/edit` endpoint has no regressions:

1. Insert an image into a post
2. Click on the crop symbol in the toolbar of the image block
3. Edit and save the image. 
4. Check in the Network tab of your browser's console that the POST to `media/{$id}/edit` took place with the correct payload.
5. Check that a new file has been created in the Media Library `/wp-admin/upload.php`

<img width="645" height="305" alt="Screenshot 2025-09-24 at 2 22 41 pm" src="https://github.com/user-attachments/assets/a1a1c4d2-b7d0-46af-b463-019f82945374" />

<img width="737" height="146" alt="Screenshot 2025-09-24 at 2 24 50 pm" src="https://github.com/user-attachments/assets/bad60178-62a0-471a-83ab-0f73b7bf7022" />

### Media processing experiment smoke testing

If you have time, smoke test the media processing experiment since this PR changes a few of the hooks and class names:

(Taken from https://github.com/WordPress/gutenberg/pull/70623)

1. Ensure experiments->client side media is enabled
2. Upload an image in the block editor
3. Monitor the network panel for activity as each image is updated. Check the files saved to the server and compare them to the files generated when the feature is not enabled.

ramonjd committed:
Enhance Gutenberg REST API for attachments: Add new controller and override existing endpoint

- Introduced `Gutenberg_Media_Processing_REST_Attachments_Controller` to manage attachment endpoints.
- Updated `gutenberg_filter_attachment_post_type_args` to use the new controller.
- Added functionality to override the REST controller for the `attachment` post type.
- Removed the deprecated `Gutenberg_REST_Attachments_Controller` and its associated tests.

This change improves the handling of media processing in the Gutenberg editor.

ramonjd committed:
Refactor Gutenberg REST Attachments Controller: Improve code formatting and consistency

- Updated PHP files to ensure consistent indentation and formatting in `Gutenberg_REST_Attachments_Controller`, `rest-api.php`, and associated test files.
- Enhanced readability by aligning comments and code structure.
- Ensured all files end with a newline for better compatibility.

These changes contribute to cleaner code and maintainability in the Gutenberg REST API for attachments.

ramonjd committed:
Update PHPCS configuration and enhance mock image editor documentation

- Added exclusion for `mock-image-editor.php` in `phpcs.xml.dist` to prevent linting errors.
- Enhanced documentation in `mock-image-editor.php` to clarify its purpose and usage in tests, including a reference to the corresponding class in WordPress Core.

These changes improve code quality and maintainability in the testing environment.

ramonjd committed:
Refactor attachment post type arguments: Simplify function signature and update filter

- Removed the unused `$post_type` parameter from `gutenberg_filter_attachment_post_type_args` function.
- Updated the `add_filter` call to reflect the change in the number of accepted arguments.

These changes enhance code clarity and maintainability in the Gutenberg media processing functionality.

ramonjd commented on lib/experimental/media/class-gutenberg-media-processing-rest-attachments-controller.php:
```diff
@@ -1,14 +1,14 @@
+ * Class Gutenberg_Media_Processing_REST_Attachments_Controller.
  */
-class Gutenberg_REST_Attachments_Controller extends WP_REST_Attachments_Controller {
+class Gutenberg_Media_Processing_REST_Attachments_Controller extends Gutenberg_REST_Attachments_Controller {
```
Extending non-experimental attachment class to avoid naming conflicts.

ramonjd commented on phpunit/includes/mock-image-editor.php:
```diff
@@ -0,0 +1,86 @@
+<?php
+
+/**
+ * Mock image editor for testing.
```
Required to mock the image editor class.

ramonjd commented on lib/experimental/media/load.php:
```diff
@@ -85,26 +85,21 @@ function gutenberg_media_processing_filter_rest_index( WP_REST_Response $respons
 	return $args;
 }
 
-add_filter( 'register_post_type_args', 'gutenberg_filter_attachment_post_type_args', 10, 2 );
```
The register_attachment_post_type_args filter is more specific and runs after the general register_post_type_args filter.

This ensures the integrity of the inheritance chain:

```
WP_REST_Attachments_Controller
    ↓
Gutenberg_REST_Attachments_Controller (WordPress 6.9 compatibility)
    ↓
Gutenberg_Media_Processing_REST_Attachments_Controller (Media experiment)
```

cc @swissspidy or @adamsilverstein for a confidence check and whether you needed `register_post_type_args` specifically. I smoke tested the media processing experiment and couldn't find any issues.

ramonjd commented on phpunit/experimental/media/media-processing-test.php:
```diff
@@ -55,16 +55,12 @@ public function test_get_all_image_sizes() {
+			array( 'rest_controller_class' => Gutenberg_Media_Processing_REST_Attachments_Controller::class ),
 			gutenberg_filter_attachment_post_type_args( array(), 'attachment' )
 		);
-		$this->assertSame(
-			array(),
-			gutenberg_filter_attachment_post_type_args( array(), 'post' )
-		);
```
No longer required due to switch to `register_attachment_post_type_args`

ramonjd commented on lib/compat/wordpress-6.9/class-gutenberg-rest-attachments-controller.php:
```diff
@@ -0,0 +1,506 @@
+		foreach ( $modifiers as $modifier ) {
+			$args = $modifier['args'];
+			switch ( $modifier['type'] ) {
+				case 'flip':
+					/*
+					 * Flips the current image.
+					 * The vertical flip is the first argument (flip along horizontal axis), the horizontal flip is the second argument (flip along vertical axis).
+					 * See: WP_Image_Editor::flip()
+					 */
+					$result = $image_editor->flip( 0 !== (int) $args['flip']['vertical'], 0 !== (int) $args['flip']['horizontal'] );
+					if ( is_wp_error( $result ) ) {
+						return new WP_Error(
+							'rest_image_flip_failed',
+							__( 'Unable to flip this image.', 'gutenberg' ),
+							array( 'status' => 500 )
+						);
+					}
+					break;
```
Flips image.

ramonjd commented on lib/compat/wordpress-6.9/class-gutenberg-rest-attachments-controller.php:
```diff
@@ -0,0 +1,506 @@
+		$original_attachment_post = get_post( $attachment_id );
+
+		// Check request fields and assign default values.
+		$new_attachment_post                 = $this->prepare_item_for_database( $request );
```
Sets and sanitizes any incoming caption, description, and title, post and alt_text fields

andrewserong commented on lib/compat/wordpress-6.9/class-gutenberg-rest-attachments-controller.php:
```diff
@@ -0,0 +1,506 @@
+		} else {
+			$modifiers = array();
+
+			if ( ! empty( $request['flip'] ) ) {
```
I noticed that further down we're expecting both the horizontal and vertical keys to be available within `flip`.
Do we need to check that `horizontal` and `vertical` are set within it? I.e. something like the following rather than a `! empty` check?:

```suggestion
			if ( isset( $request['flip']['horizontal'], $request['flip']['vertical'] ) ) {
```

ramonjd replied:
Good idea. I'm not sure we need to check for both. I think it can be either or technically. 

In that case I probably need to then tighten up the flip operation below

> $result = $image_editor->flip( 0 !== (int) $args['flip']['vertical'], 0 !== (int) $args['flip']['horizontal'] );

andrewserong replied:
> Good idea. I'm not sure we need to check for both. I think it can be either or technically.

Ah, true! Yeah, whichever way works 👍

ramonjd replied:
Looks like the Image Classes don't require both, e.g., https://github.com/WordPress/wordpress-develop/blob/6.8.2/src/wp-includes/class-wp-image-editor-imagick.php#L761


I'll update this change and add a couple more tests.

Thanks!

ramonjd committed:
Enhance image flipping functionality in Gutenberg REST Attachments Controller

- Updated the `Gutenberg_REST_Attachments_Controller` to handle both vertical and horizontal flipping of images by modifying the request structure.
- Refactored the test case to reflect the new functionality, adding a separate test for vertical flipping.

swissspidy commented:
Since this doesn't affect any functionality in Gutenberg, I don't see why the PR is made against this repo. I strongly suggest making a PR for WP core directly.

ramonjd commented:
> Since this doesn't affect any functionality in Gutenberg, I don't see why the PR is made against this repo.

Yet.

It's here for the case where follow up experiments, that rely on these changes, are added to the plugin. There are a couple of things we'd like to try so best to be flexible.

ramonjd commented:
Closing in favor of https://github.com/WordPress/wordpress-develop/pull/10041
